File: dumper/luarun.js
import { execFile } from 'node:child_process';
import crypto from 'node:crypto';
import fs from 'node:fs/promises';
import os from 'node:os';
import path from 'node:path';
import { promisify } from 'node:util';

const execFileAsync = promisify(execFile);

export class LuaRunner {
  constructor() {
    // Generate a unique temporary directory
    this.tmp = fs.mkdtemp(path.join(os.tmpdir(), 'lrun_'));
    this.cache = {};
  }

  /**
   * Constructs the Lua execution harness.
   */
  buildHarness(code, tag = 'analysis') {
    const tpl = `
local _tag = "${tag}"
local _t0 = os.clock()

local function _log(kind, msg)
    print(string.format("[%s] %s", kind, msg))
end

local _env = {}
setmetatable(_env, {
    __index = function(t, k)
        if k == "print" then
            return function(...)
                local parts = {...}
                _log("OUT", table.concat(parts, "\\t"))
            end
        end
        return nil
    end
})

${code}

_log("DONE", string.format("%.3fs", os.clock() - _t0))
`;
    return tpl;
  }

  /**
   * Executes the Lua code using the host system's lua/luau binary.
   */
  async execLua(code) {
    const tmpDir = await this.tmp;
    const fpath = path.join(tmpDir, 'run.lua');

    await fs.writeFile(fpath, code, 'utf-8');

    try {
      // Use execFile to run the Lua CLI safely with a 10s timeout
      const { stdout, stderr } = await execFileAsync('lua', [fpath], {
        timeout: 10000,
        maxBuffer: 1024 * 1024 * 10, // 10MB buffer limit
      });

      return {
        ok: true,
        out: stdout || '',
        err: stderr || '',
        code: 0,
      };
    } catch (err) {
      if (err.killed && err.signal === 'SIGTERM') {
        return {
          ok: false,
          out: '',
          err: 'Lua execution timeout (10s)',
          code: -1,
        };
      }
      if (err.code === 'ENOENT') {
        return {
          ok: false,
          out: '',
          err: 'Lua interpreter not found',
          code: -1,
        };
      }
      return {
        ok: false,
        out: err.stdout || '',
        err: err.stderr || err.message || String(err),
        code: typeof err.code === 'number' ? err.code : -1,
      };
    }
  }

  /**
   * Scans function signatures via Regex.
   */
  getFuncs(src) {
    const pats = [
      /function\s+(\w+)\(([^)]*)\)/g,
      /local\s+function\s+(\w+)\(([^)]*)\)/g,
      /(\w+)\s*=\s*function\(([^)]*)\)/g,
    ];
    const out = [];

    for (const pat of pats) {
      const matches = src.matchAll(pat);
      for (const m of matches) {
        const rawParams = m[2] ? m[2].trim() : '';
        out.push({
          name: m[1],
          params: rawParams ? rawParams.split(',').map((p) => p.trim()) : [],
          sig: m[0],
        });
      }
    }
    return out;
  }

  /**
   * Calculates string and hash metrics for the source code.
   */
  metrics(src) {
    const lines = src.split('\n');
    const totalChars = lines.reduce((acc, l) => acc + l.length, 0);

    return {
      lines: lines.length,
      chars: src.length,
      funcs: this.getFuncs(src).length,
      md5: crypto.createHash('md5').update(src).digest('hex'),
      sha256: crypto.createHash('sha256').update(src).digest('hex'),
      avg_len: totalChars / Math.max(lines.length, 1),
    };
  }

  /**
   * Processes a target file and caches results.
   */
  async process(filePath) {
    let src;
    try {
      src = await fs.readFile(filePath, { encoding: 'utf-8' });
    } catch (e) {
      return {
        file: filePath,
        error: `Failed to read file: ${e.message}`,
        metrics: null,
        functions: [],
        run: null,
      };
    }

    const m = this.metrics(src);
    const harness = this.buildHarness(src, 'full');
    const run = await this.execLua(harness);

    const result = {
      file: filePath,
      metrics: m,
      functions: this.getFuncs(src),
      run: run,
    };

    this.cache[filePath] = result;
    return result;
  }

  /**
   * Dumps accumulated cache results into a JSON file.
   */
  async dump(outPath) {
    const report = {
      version: '2.0.0',
      files: Object.keys(this.cache).length,
      results: this.cache,
    };

    await fs.writeFile(outPath, JSON.stringify(report, null, 2), 'utf-8');
    return outPath;
  }
}

File: dumper/luarun.py
import subprocess
import tempfile
import os
import json
import hashlib
import re
from typing import Dict, List


class LuaRunner:
    def __init__(self):
        self.tmp = tempfile.mkdtemp(prefix="lrun_")
        self.cache = {}

    def build_harness(self, code: str, tag: str = "analysis") -> str:
        tpl = """
local _tag = "{tag}"
local _t0 = os.clock()

local function _log(kind, msg)
    print(string.format("[%s] %s", kind, msg))
end

local _env = {{}}
setmetatable(_env, {{
    __index = function(t, k)
        if k == "print" then
            return function(...)
                local parts = {{...}}
                _log("OUT", table.concat(parts, "\\t"))
            end
        end
        return nil
    end
}})

{code}

_log("DONE", string.format("%.3fs", os.clock() - _t0))
"""
        return tpl.format(tag=tag, code=code)

    def exec_lua(self, code: str) -> Dict:
        fpath = os.path.join(self.tmp, "run.lua")
        with open(fpath, 'w', encoding='utf-8') as f:
            f.write(code)

        try:
            r = subprocess.run(
                ['lua', fpath],
                capture_output=True,
                text=True,
                timeout=10
            )
            return {
                'ok': r.returncode == 0,
                'out': r.stdout if r.stdout else '',
                'err': r.stderr if r.stderr else '',
                'code': r.returncode,
            }
        except subprocess.TimeoutExpired:
            return {'ok': False, 'out': '', 'err': 'Lua execution timeout (10s)', 'code': -1}
        except FileNotFoundError:
            return {'ok': False, 'out': '', 'err': 'Lua interpreter not found', 'code': -1}
        except Exception as e:
            return {'ok': False, 'out': '', 'err': str(e), 'code': -1}

    def get_funcs(self, src: str) -> List[Dict]:
        pats = [
            r'function\s+(\w+)\(([^)]*)\)',
            r'local\s+function\s+(\w+)\(([^)]*)\)',
            r'(\w+)\s*=\s*function\(([^)]*)\)',
        ]
        out = []
        for pat in pats:
            for m in re.finditer(pat, src):
                out.append({
                    'name': m.group(1),
                    'params': [p.strip() for p in m.group(2).split(',')] if m.group(2).strip() else [],
                    'sig': m.group(0),
                })
        return out

    def metrics(self, src: str) -> Dict:
        lines = src.split('\n')
        return {
            'lines': len(lines),
            'chars': len(src),
            'funcs': len(self.get_funcs(src)),
            'md5': hashlib.md5(src.encode()).hexdigest(),
            'sha256': hashlib.sha256(src.encode()).hexdigest(),
            'avg_len': sum(len(l) for l in lines) / max(len(lines), 1),
        }

    def process(self, path: str) -> Dict:
        try:
            with open(path, 'r', encoding='utf-8', errors='ignore') as f:
                src = f.read()
        except Exception as e:
            return {
                'file': path,
                'error': f'Failed to read file: {str(e)}',
                'metrics': None,
                'functions': [],
                'run': None,
            }

        m = self.metrics(src)
        harness = self.build_harness(src, "full")
        run = self.exec_lua(harness)

        result = {
            'file': path,
            'metrics': m,
            'functions': self.get_funcs(src),
            'run': run,
        }

        self.cache[path] = result
        return result

    def dump(self, out_path: str):
        report = {
            'version': '2.0.0',
            'files': len(self.cache),
            'results': self.cache,
        }
        with open(out_path, 'w', encoding='utf-8') as f:
            json.dump(report, f, indent=2, ensure_ascii=False)
        return out_path
